rfcs/rdm-0045-versioning-support.md

```diff
@@ -0,0 +1,530 @@
+
+## Backend Design
+
+The backend consist of the data access, service and presentation layers.
+
+### Data access layer
+
+#### Version relationships
+
+We propose that the information previously held by Invenio-PIDRelations is moved directly to the record database table. The current attributes of the ``RDMRecordMetadata`` model includes:
+
+- ``id `` - UUID of the record
+- ``json`` - Record JSON document
+- ``version_id`` - Version counter used for optimistic concurrency control.
+- ``created `` - Creation timestamp.
+- ``updated`` - Modification timestamp.
+- ``bucket_id`` - Foreign key (UUID) to a Files-REST bucket.
+
+We propose to add the following two attributes:
+
+- ``parent_id`` - Foreign key (UUID) to a Concept Record (a new entity).
+- ``parent_order`` - Integer (4 bytes) defining the order of siblings.
+- ``parent_latest`` - Boolean to indicate if the record is the latest version.
+
+The name ``parent`` could alternatively be ``concept``, but we would like to avoid naming that conflicts with the ``version_id``. Technically ``version_id`` is the **revision identifier** however, due to history reasons and API this is not an easy rename to perform.
+
+In addition a database index should be defined to allow for easily querying all records related to a concept record in the correct order (thus likely a combined index on ``parent_id`` and ``parent_order``):
+
+```SQL
+SELECT * FROM rdm_records_metadata WHERE parent_id=... ORDER BY parent_order
+```
+
+#### Concept Record
+
+We also propose to add a new entity, a ``ConceptRecord``, which represents the concept record and are able to hold properties for all records such as ownership and permissions.
+
+The model (table name ``rdm_records_concept``) would be like a usual record with the following database columns:
+
+- ``id `` - UUID of the concept record
+- ``json`` - Record JSON document
+- ``version_id`` - Version counter used for optimistic concurrency control.
+- ``created `` - Creation timestamp.
+- ``updated`` - Modification timestamp.
+
+The concept record would hold JSON documents similar to below:
+
+```json
+{
+    "id": "1234-abcde",
+    "pid": {
+        ...
+    },
+    "access": {
+        "owned_by": [{"user": 1}],
+        "grants: [
+            ...
+        ]
+    },
+}
+```
+
+Essentially the concept persistent identifier, the access ownership and access grants would be moved from the existing record to the new concept record.
+
+**Why a new entity?**
+
+The new model allows keeping the ownership and permissions separate from the single version records. Alternatively, we would have to duplicate (and update) access information in multiple specific version records.
+
+Also, having just a single concept record for both records and drafts, means that access information can be updated not only from the deposit form, but also e.g. share button on the landing page record.
```

Comment:
(it's really the fact that drafts and recors share the same recid that makes it easy to have share button on each :wink: ). Having a single concept record for both record and drafts is "natural":

```
Concept Record
|
|___ Record V2
|___ Record V1 ------ Draft V1
|___ Record V0 
```